server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -99,95 +206,219 @@ public DeferringBucketCollector getDeferringCollector() {
-                    LongHash newBucketOrds = new LongHash(1, context.bigArrays());
-                    long[] mergeMap = new long[(int) oldBucketOrds.size()];
-                    preparedRounding = roundingPreparer.apply(roundingInfos[++roundingIdx].rounding);
-                    for (int i = 0; i < oldBucketOrds.size(); i++) {
-                        long oldKey = oldBucketOrds.get(i);
-                        long newKey = preparedRounding.round(oldKey);
-                        long newBucketOrd = newBucketOrds.add(newKey);
-                        if (newBucketOrd >= 0) {
-                            mergeMap[i] = newBucketOrd;
-                        } else {
-                            mergeMap[i] = -1 - newBucketOrd;
-                        }
-                    }
-                    mergeBuckets(mergeMap, newBucketOrds.size());
-                    if (deferringCollector != null) {
-                        deferringCollector.mergeBuckets(mergeMap);
+            private byte collectValue(LeafBucketCollector sub, long owningBucketOrd, byte roundingIdx, int doc, long rounded)
+                    throws IOException {
+                long bucketOrd = bucketOrds.add(owningBucketOrd, rounded);
+                if (bucketOrd < 0) { // already seen
+                    bucketOrd = -1 - bucketOrd;
+                    collectExistingBucket(sub, doc, bucketOrd);
+                    return roundingIdx;
+                }
+                collectBucket(sub, doc, bucketOrd);
+                liveBucketCountUnderestimate = context.bigArrays().grow(liveBucketCountUnderestimate, owningBucketOrd + 1);
+                int estimatedBucketCount = liveBucketCountUnderestimate.increment(owningBucketOrd, 1);
+                return increaseRounding(owningBucketOrd, estimatedBucketCount, rounded, roundingIdx);
+            }
+
+            /**
+             * Increase the rounding of {@code owningBucketOrd} using
+             * estimated, bucket counts, {@link #rebucket() rebucketing} the all
+             * buckets if the estimated number of wasted buckets is too high.
+             */
+            private byte increaseRounding(long owningBucketOrd, int oldEstimatedBucketCount, long newKey, byte oldRounding) {
+                if (oldRounding >= roundingInfos.length - 1) {
+                    return oldRounding;
+                }
+                if (bounds.size() < owningBucketOrd * 2 + 2) {
+                    long oldSize = bounds.size();
+                    bounds = context.bigArrays().grow(bounds, owningBucketOrd * 2 + 2);
+                    for (long b = oldSize; b < bounds.size(); b++) {
+                        bounds.set(b, (b & 1L) == 0L ? Long.MAX_VALUE : Long.MIN_VALUE);
                     }
-                    bucketOrds = newBucketOrds;
                 }
+                long min = Math.min(bounds.get(owningBucketOrd * 2), newKey);
+                bounds.set(owningBucketOrd * 2, min);
```

Comment:
I think we can move these two `bounds.set()` inside the `if() { ... }` down below, to avoid unnecessary setting when we have to fall through and increase the rounding.  Not that it's terribly expensive, but if we can avoid always good :)